Libraries/LibGUI/EditingEngine.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/Noncopyable.h>
+#include <LibGUI/Event.h>
+#include <LibGUI/TextDocument.h>
+
+namespace GUI {
+
+enum CursorWidth {
+    NARROW,
+    WIDE
+};
+
+class EditingEngine {
+    AK_MAKE_NONCOPYABLE(EditingEngine);
+    AK_MAKE_NONMOVABLE(EditingEngine);
+
+public:
+    virtual ~EditingEngine();
+
+    virtual CursorWidth cursor_width() const { return NARROW; }
+
+    void attach(TextEditor& editor);
+    void detach();
```

Comment:
Not gonna lie, this wasn't exactly what I had in mind when I suggested moving the vim logic to a library...what I had in mind was more of an engine that had its own set of commands (e.g. `MoveOneLeft`, `MoveOneRight`, `MoveUpSteps`, etc.) and had zero ties to any specific editor implementation - that's why I gave neovim as an example, but it seems I was ambiguous at best.

Reply:
Such decoupling would be quite hard to implement since the interaction surface is big. For instance, handling selections will require relatively tight integration. Is this worth the effort at this initial step?

Reply:
I'm not entirely sure, if you think it won't hurt a future refactoring too much, you can leave it as it is.

(possibly for the future)
regarding the selections, it really does look like everything depends on the selection, maybe the engine should have its own idea of the selection, and the editor would just show said selection?

Reply:
> maybe the engine should have its own idea of the selection, and the editor would just show said selection?

I've thought about this, but it would require that mouse selection be handled by this module as well (or at least, forwarded there, from the text editor). Which could be very well be the case, but I thought it might be too big of a refactoring, just to add a vim emulation (at least in this step).

Reply:
> mouse
> vim mode

disable that lol
(just a joke ^)

Reply:
On a more serious note, it's alright to have it be dependent on TextEditor for now, we can always change it later.